src/pages/ReimbursementAccount/ACHContractStep.js

```diff
@@ -233,18 +259,27 @@ class ACHContractStep extends React.Component {
                                     <IdentityForm
                                         translate={this.props.translate}
                                         style={[styles.mb2]}
-                                        onFieldChange={values => this.clearErrorAndSetBeneficialOwnerValues(index, values)}
-                                        values={{
-                                            firstName: owner.firstName || '',
-                                            lastName: owner.lastName || '',
-                                            street: owner.street || '',
-                                            city: owner.city || '',
-                                            state: owner.state || '',
-                                            zipCode: owner.zipCode || '',
-                                            dob: owner.dob || '',
-                                            ssnLast4: owner.ssnLast4 || '',
+                                        defaultValues={{
+                                            firstName: this.getDefaultStateForField('firstName'),
+                                            lastName: this.getDefaultStateForField('lastName'),
+                                            street: this.getDefaultStateForField('beneficialOwnerAddressStreet'),
+                                            city: this.getDefaultStateForField('beneficialOwnerAddressCity'),
+                                            state: this.getDefaultStateForField('beneficialOwnerAddressState'),
+                                            zipCode: this.getDefaultStateForField('beneficialOwnerAddressZipCode'),
+                                            dob: this.getDefaultStateForField('dob'),
+                                            ssnLast4: this.getDefaultStateForField('ssnLast4'),
+                                        }}
+                                        inputKeys={{
+                                            firstName: 'firstName',
+                                            lastName: 'lastName',
+                                            dob: 'dob',
+                                            ssnLast4: 'ssnLast4',
+                                            street: 'beneficialOwnerAddressStreet',
+                                            city: 'beneficialOwnerAddressCity',
+                                            state: 'beneficialOwnerAddressState',
+                                            zipCode: 'beneficialOwnerAddressZipCode',
```

Comment:
I'll look into a solution for these since we might need to make changes to the Form component itself

Reply:
@grgia I took a look at dynamically adding/removing `IdentityForm` and the solution below should work:

1. Store `beneficialOwners` in local state like we do now. However, we'll only store an ID for the `IdentityForm` instead of the full form input keys.
2. Update `addBeneficialOwner` to store this ID in `beneficialOwners` and set the Form draft values accordingly. Something like:

```
addBeneficialOwner() {
    this.setState((prevState) => {
        const beneficialOwners = [...prevState.beneficialOwners, NumberUtils.rand64()];

        // We set 'beneficialOwners' to null first because we don't have a way yet to replace a specific property without merging it.
        // We don't use the debounced function because we want to make both function calls.
        FormActions.setDraftValues(ONYXKEYS.REIMBURSEMENT_ACCOUNT, {beneficialOwners: null});
        FormActions.setDraftValues(ONYXKEYS.REIMBURSEMENT_ACCOUNT, {beneficialOwners});

        return {beneficialOwners};
    });
}
```
3. Update `removeBeneficialOwner` to take the ID of the `IdentityForm` we are removing and update the Form draft accordingly. Like so:

```
removeBeneficialOwner(beneficialOwner) {
    this.setState((prevState) => {
        const beneficialOwners = _.without(prevState.beneficialOwners, beneficialOwner);

        // We set 'beneficialOwners' to null first because we don't have a way yet to replace a specific property without merging it.
        // We don't use the debounced function because we want to make both function calls.
        FormActions.setDraftValues(ONYXKEYS.REIMBURSEMENT_ACCOUNT, {beneficialOwners: null});
        FormActions.setDraftValues(ONYXKEYS.REIMBURSEMENT_ACCOUNT, {beneficialOwners});

        return {beneficialOwners};
    });
}
```

4. In the render method, map over the `beneficialOwners` state, which now stores IDs, and pass a key in the format `beneficialOwner.${id}.firstName` to both `defaultValues` and `inputKeys`. Something like:

```
defaultValues={{
    firstName: ReimbursementAccountUtils.getDefaultStateForField(this.props, `beneficialOwner.${id}.firstName`),
    ...
    }}
inputKeys={{
    firstName: `beneficialOwner.${id}.firstName`,
    ...
}}
```
5. Add `onValueChange={value => this.setState({hasOtherBeneficialOwners: value})}` to the `hasOtherBeneficialOwners` checkbox.
6. Update the call to `this.removeBeneficialOwner` to pass the ID of the `IdentityForm` we are removing.
7. Update the `validate` function to use the new keys.

```
_.each(this.state.beneficialOwners, (id) => {
    if (!ValidationUtils.isRequiredFulfilled(values[`beneficialOwner.${id}.firstName`])) {
        errors[`beneficialOwner.${id}.firstName`] = this.props.translate('bankAccount.error.firstName');
    }
    ...
```

8. The main issue with this approach is that it doesn't update the state in Form, so it could still pass data stored for deleted `IdentityForms` (if the user removed a beneficial owner as they are editing the form). The only drawback that I can think of is that we will have to filter these values in the `submit` function so we don't send them with the API request. I think this is a fine trade off considering that it makes the rest of the implementation easier.

Let me know what you think of this approach!

Reply:
This method looks like it's working well! I wrote out my plan for testing, so I still need to QA edge cases

Reply:
Sounds good! I'll take another look today! Thanks for working on this.

Reply:
Was there some reason we could not just use an index here? 

If not, I'd maybe update the language to say `key` instead of `id` since I normally associated users with `accountID`. Would also leave some kind of comment so it's clear why we are doing this in the code.

Reply:
>Was there some reason we could not just use an index here?

Scenario: User adds 2 beneficial owners, but then removes the 1st one.

Right now we can only update values stored in Form, but not completely remove them. So if we used an index, when the user removes the 1st beneficial owner, we'd have to update all the values in the 1st index to the values stored in the 2nd index and then set the 2nd to empty strings.

Another option here would be to add a method to Form to allow removing values completely. We'd still have to pass an index though, so IMO passing a random key is simpler.

Reply:
>if we used an index, when the user removes the 1st beneficial owner, we'd have to update all the values in the 1st index to the values stored in the 2nd index and then set the 2nd to empty strings.

I think I'm not getting it still - maybe you can break it down some more? 

I want to say -- what if the `beneficialOwners` form data remains an array and not an object keyed by an `id`? And when we need to update it we just completely replace the array with a new one?

Reply:
>I think I'm not getting it still - maybe you can break it down some more?

Form keeps all its data in an object keyed by `inputIDs`. If we use a beneficial owner index, form would have something like this stored:

```
{
   beneficialOwner.1.firstName: "Jane"
   beneficialOwner.1.lastName: "Doe"
   beneficialOwner.2.firstName: "John"
   beneficialOwner.2.lastName: "Doe"
}
```

If we remove `Jane Doe` we would have to update form data accordingly, i.e. `beneficialOwner.1.firstName = beneficialOwner.2.firstName` and remove the keys for `beneficialOwner.2`

I hope this made things clearer and not more confusing 😅 

>what if the beneficialOwners form data remains an array and not an object keyed by an id?
And when we need to update it we just completely replace the array with a new one?

This would be ideal, but how can we achieve this?

Reply:
>I hope this made things clearer and not more confusing 😅

I think you kind of explain the parts that I am familiar with, but not the parts that I'm not so the problem is still unclear.

>If we remove Jane Doe we would have to update form data accordingly, i.e. beneficialOwner.1.firstName = beneficialOwner.2.firstName and remove the keys for beneficialOwner.2

Yes, can we not store the form data as an array?

Reply:
>Yes, can we not store the form data as an array?

AFAIK we can't in the current Form implementation. We store whatever data type the child input passes to the `onInputChange` callback, i.e. string, int, bools, but I don't think any type of input returns arrays.

Reply:
OK I am still a little confused, but haven't yet given this a proper investigation to produce any alternatives. Something does not sit right with me about the random id thing. I feel like I would have remembered this from the Design Doc (maybe it didn't come up). It has the feeling of something we might want to bring up in Slack, decide together the best way to do it, and then update everyone on the process (maybe modify the very awesome `FORMS.md` doc 😉).

Reply:
I don't think this came up in the doc (at least I don't recall it either). IMO the random id solution solves this issue quite well, but I'm open to discussing this further in Slack if others prefer that.

Reply:
so, sounds like we agree that it's a good thing to discuss and that you are offering to lead a discussion on it? Or no? 😄 

It might feel like a small battle - but I am concerned with this solution because it feels like something I would never remember how to do or know the reason why a random ID is used. I am _barely_ understanding the explanation about why it is needed. I'd say that's a sign we haven't solved it in the best way possible (maybe I'm wrong). But if it has to be difficult to understand then at least we can document it better.